src/Workspaces/Core/Portable/Workspace/Solution/Solution.cs

```diff
@@ -1145,17 +1264,29 @@ public Solution WithDocumentSourceCodeKind(DocumentId documentId, SourceCodeKind
         /// </summary>
         public Solution WithDocumentTextLoader(DocumentId documentId, TextLoader loader, PreservationMode mode)
         {
-            return WithDocumentTextLoader(documentId, loader, text: null, mode: mode);
-        }
+            CheckContainsDocument(documentId);
 
-        internal Solution WithDocumentTextLoader(DocumentId documentId, TextLoader loader, SourceText? text, PreservationMode mode)
-        {
-            var newState = _state.WithDocumentTextLoader(documentId, loader, text, mode);
-            if (newState == _state)
+            if (loader == null)
             {
-                return this;
+                throw new ArgumentNullException(nameof(loader));
             }
 
+            if (!mode.IsValid())
+            {
+                throw new ArgumentOutOfRangeException(nameof(mode));
+            }
+
+            return UpdateDocumentTextLoader(documentId, loader, text: null, mode: mode);
+        }
+
+        internal Solution UpdateDocumentTextLoader(DocumentId documentId, TextLoader loader, SourceText? text, PreservationMode mode)
+        {
+            var newState = _state.UpdateDocumentTextLoader(documentId, loader, text, mode);
+
+            // Note: state is currently not reused.
+            // If UpdateDocumentTextLoader is changed to reuse the state replace this assert with Solution instance reusal.
```

Comment:
Thanks for the comment. My worry with asserts is they are often used to state not how the system works today but how it _should_ work, so any change that breaks an assert is "wrong". This is most definitely a "this is just how it works, but if you fix it we can then assert something better" so great.